adaptdl/adaptdl/torch/parallel.py

```diff
@@ -59,9 +67,12 @@ def __init__(self, model, optimizer, lr_scheduler=None, mp_scaler=None,
         for param in model.parameters():
             param.register_hook(functools.partial(self._backward_hook, param))
 
-        # Setup for AdaScale, must be after registering backward hooks!
-        self.adascale = AdaScale(self, optimizer, mp_scaler=mp_scaler,
-                                 patch_optimizer=True)
+        # Setup for the scaling_rule, must be after registering backward hooks
+        # because some of them need to register their own backward hooks.
+        self.gns = GradientNoiseScale(self, optimizer, mp_scaler=mp_scaler)
+        self.scaling_rule = scaling_rule or AdaScale()
```

Comment:
I think this would be easier to use with a string to specify the class (and then create the object here), rather than forcing the user to create the object themselves

Reply:
Unless having the users be able to extend the scaling rules themselves is important

Reply:
yeah, I was using string arg and then Aurick and I had a discussion here. The interface we currently have does allow for more flexibility. And for some scaling rules, like LEGW, we also need additional args to be passed in for initialization.
https://github.com/petuum/adaptdl/pull/76#discussion_r546293108